src/saltext/vmware/utils/license_mgr.py

```diff
@@ -0,0 +1,448 @@
+
+    Returns:
+        True - if successful, license removed
+    """
+    lic_mgr = get_license_mgr(service_instance)
+    if not lic_mgr:
+        log.debug(
+            f"attempting to remove license '{license_key}' but unable to find license manager"
+        )
+        return False
+
+    lic_keys = _list_lic_keys(lic_mgr.licenses)
+    log.debug(
+        f"attempting to remove license '{license_key}' from list of existing license keys '{lic_keys}'"
+    )
+    if license_key not in lic_keys:
+        log.debug(
+            f"cannot remove license '{license_key}' since not found in list of existing license keys '{lic_keys}'"
+        )
+        return False
+
+    # Need to fine where license key is assigned and then un-assign it,
+    # and only then can the license actually be removed, that is, RemoveLicense is a noop
+    # if the license is assigned
+    #
+    # RemoveAssignedLicense takes entityId, reason to find were license key is assigned
+    # Note: an entity can only have one license assigned
+
+    try:
+        # need to extract entity this license is intended to be removed from
+        # choices are:
+        #   1 vCenter - entity_id = service_instance.content.about.instanceUuid
+        #   2 ESXi - entity_id = esxi host _moId
+        #   3 cluster - entity_id = cluster _moId
+
+        entity_id = None  # TBD miracle happens here and have value
+        datacenter_ref = None
+        if datacenter_name or cluster_name or esxi_hostname:
+            if datacenter_name:
+                # need to get named datacenter's reference
+                datacenter_ref = utils_datacenter.get_datacenter(service_instance, datacenter_name)
+                log.debug(
+                    f"retrieved datacenter ref '{datacenter_ref }' for datacenter '{datacenter_name}'"
+                )
+
+            if cluster_name:
+                # need to get named cluster's reference
+                cluster_ref = utils_cluster.get_cluster(dc_ref=datacenter_ref, cluster=cluster_name)
+                entity_id = cluster_ref._moId
+                log.debug(
+                    f"retrieved entityId '{entity_id}' from cluster ref '{cluster_ref }' for cluster '{cluster_name}' and datacenter '{datacenter_name}'"
+                )
+
+            if esxi_hostname:
+                # need to get named esxi server
+                esxi_hosts = utils_esxi.get_hosts(service_instance, datacenter_name, esxi_hostname)
+
+                # returns hosts list of dicts
+                if not esxi_hosts:
+                    log.debug(f"Failed to find esxi hostname '{esxi_hostname}'")
+                    return False
+
+                if len(esxi_hosts) > 1:
+                    log.error(
+                        f"Failed, found multiple instances of esxi hostname '{esxi_hostname}', hosts returned '{esxi_hosts}'"
+                    )
+                    return False
+
+                entity_id = esxi_hosts[0]._moId
```

Comment:
See line 379, datacenter_ref is used to get cluster_ref.  Style is somewhat similar to how Ansible did things btw.
Have to check Datacenter to get Cluster information.
Could change to 
```
if datacenter_name or cluster_name:
  xxxxxxx
elif esxi_hostname:
  xxxxxxx
```
But again datacenter_name is used with esxi_hostname in line 387, so don't think it matters

Reply:
Ahhhh... I see. I think it might actually make more sense to flip the condition then:

```
        if not any((datacenter_name, cluster_name, esxi_hostname)):
            srv_content = get_service_content(service_instance)
            entity_id = srv_content.about.instanceUuid
        else:
            if datacenter_name:
                # need to get named datacenter's reference
                datacenter_ref = utils_datacenter.get_datacenter(service_instance, datacenter_name)
                log.debug(
                    f"retrieved datacenter ref '{datacenter_ref }' for datacenter '{datacenter_name}'"
                )

            if cluster_name:
                # need to get named cluster's reference
                cluster_ref = utils_cluster.get_cluster(dc_ref=datacenter_ref, cluster=cluster_name)
                entity_id = cluster_ref._moId
                log.debug(
                    f"retrieved entityId '{entity_id}' from cluster ref '{cluster_ref }' for cluster '{cluster_name}' and datacenter '{datacenter_name}'"
                )

            if esxi_hostname:
                # need to get named esxi server
                esxi_hosts = utils_esxi.get_hosts(service_instance, datacenter_name, esxi_hostname)

                # returns hosts list of dicts
                if not esxi_hosts:
                    log.debug(f"Failed to find esxi hostname '{esxi_hostname}'")
                    return False

                if len(esxi_hosts) > 1:
                    log.error(
                        f"Failed, found multiple instances of esxi hostname '{esxi_hostname}', hosts returned '{esxi_hosts}'"
                    )
                    return False

                entity_id = esxi_hosts[0]._moId
                if "esx" not in addedLic.editionKey:
                    log.error(
                        f"Error, License '{license_key}' does not contain a suitable Edition key '{license.editionKey}' for an ESXi Server"
                    )
                    return False
```

It *might* also be a good idea to extract this block to a function like 

```
entity_id = _get_entity_id(datacenter_name, cluster_name, esxi_hostname)
```

Reply:
done